torch/distributed/_state_dict_utils.py

```diff
@@ -50,60 +60,97 @@ def _all_gather_sharded_tensor(
     return tensor
 
 
+class CompanionMismatch(Exception):
+    ...
+
+
 def _iterate_state_dict(
     iter_object: Any,
     sharded_tensor_func: Callable,
     dtensor_func: Callable,
+    tensor_func: Callable,
     *,
     pg: Optional[dist.ProcessGroup] = None,
     device: Optional[torch.device] = None,
     cpu_offload: bool = False,
+    companion_obj: Any = None,
     ranks_only: Tuple[int, ...] = tuple(),
+    type_check: bool = True,
 ) -> Dict[str, Any]:
     # TODO: should we use pytree?
     cpu_device = torch.device("cpu")
     if isinstance(iter_object, ShardedTensor):
-        ret = sharded_tensor_func(iter_object, pg, device)
+        ret = sharded_tensor_func(iter_object, pg, device, companion_obj)
     elif isinstance(iter_object, DTensor):
-        ret = dtensor_func(iter_object, pg, device)
+        ret = dtensor_func(iter_object, pg, device, companion_obj)
+    elif isinstance(iter_object, torch.Tensor):
+        ret = tensor_func(iter_object, pg, device, companion_obj)
     elif (
-        isinstance(iter_object, (torch.Tensor, int, float, str)) or iter_object is None
+        isinstance(iter_object, (int, float, str, bytes, io.BytesIO))
+        or iter_object is None
     ):
         ret = iter_object
     elif isinstance(iter_object, dict):
+        if companion_obj is None:
+            companion_obj = {k: None for k in iter_object}
+        elif not isinstance(companion_obj, dict) or set(companion_obj.keys()) != set(
+            iter_object.keys()
+        ):
+            raise CompanionMismatch()
```

Comment:
nit: we could consider adding something like:

 `f"{type(companion_obj)} {set(companion_obj.keys()) - set(iter_object.keys()) }"` to the exception to provide some insight.

Reply:
hmm, I'm not sure if the error message is going to help since we just silently ignore the CompanionMismatch error. But it doesn't hurt to print out the keys later I think.